Display child modules nested under parents

esoergel opened the pull request:
![image](https://user-images.githubusercontent.com/2367539/53749597-af05cf00-3e75-11e9-8d59-ab54db37bcb0.png)

This PR displays child modules nested below their parents.  Rearrangements preserve the nested structure, so moving a parent also moves the children.  Child modules can be rearranged under the same parent, but not promoted to top level or moved to a new parent.

Commit by commit will make the most sense.
@emord @snopoke

esoergel committed:
Merge module link templates

No substantive changes

esoergel committed:
Merge form link templates

esoergel committed:
Track module changes by UID

esoergel committed:
Indent child modules

esoergel committed:
Pull out the top level of abstraction

esoergel committed:
Pull out update function

esoergel committed:
Pull out components of update function

esoergel committed:
Nest child modules under their parents

So drag and drop works properly

esoergel committed:
Force user to fix ordering issue if one exists

This could happen if we didn't migrate the app initially, or if the user
reverted to an old version of the app where the ordering restriction didn't exist.

esoergel committed:
Alert user of orphaned child modules

esoergel committed:
Move children if moving parent

esoergel committed:
lint

esoergel commented on corehq/apps/app_manager/static/app_manager/js/app_manager.js:
```diff
@@ -234,10 +234,14 @@ hqDefine('app_manager/js/app_manager', function () {
+                initSortable($(this));
+            });
+            $('.sort-action').hide();
+        }
```
The promptToSaveOrdering will be dismissed if the user clicks outside the modal, but they won't be able to reorder the modules, since they're not initialized here.

As per the commit message, I don't expect this situation to arise often, but it could, since we currently have applications with child modules that don't follow their parent.

orangejenny replied:
I think this is a good ui compromise.

esoergel commented on corehq/apps/app_manager/static/app_manager/js/app_manager.js:
```diff
@@ -265,8 +265,13 @@ hqDefine('app_manager/js/app_manager', function () {
+                        alertText = _.template(gettext(
+                        'Menu "<%= moduleName %>" references a parent menu that no longer exists'
+                    ))({moduleName: moduleName});
+                    hqImport('hqwebapp/js/alert_user').alert_user(alertText, "danger");
```
When they navigate to the module, there's also a warning on the page.

orangejenny replied:
So this'll show a warning on every page in app manager, right? Given that there's a warning on the most relevant page and that there's a build error, I'm not convinced this is needed.

esoergel replied:
That's a good point, I'll kill it.

esoergel commented:
This PR does force some projects to reorder their modules.  There are comparatively few projects impacted by this, but the solutions team will be reaching out to those projects to "fix" their ordering prior to releasing this feature.

orangejenny commented on corehq/apps/app_manager/templates/app_manager/partials/menu/module_link.html:
```diff
@@ -0,0 +1,109 @@
+{% load xforms_extras %}
+{% load hq_shared_tags %}
+
+{# For APP BUILDER PROTOTYPE use only #}
```
this would be a nice comment to kill

orangejenny commented on corehq/apps/app_manager/models.py:
```diff
@@ -5410,14 +5410,14 @@ def rearrange_modules(self, i, j):
         self.modules = modules
 
-    def rearrange_forms(self, to_module_id, from_module_id, i, j):
+    def rearrange_forms(self, to_module_uid, from_module_uid, i, j):
```
hooray

esoergel replied:
I did also consider switching rearrangement to work by UUID, but figured that'd be a bit too much scope creep and wouldn't really make what I'm working on any easier.  Would be nice to do though - instead of saying "rearrange from position x to position y", say "move module X to immediately after module y".  Then even if someone moved things around after you loaded the page, it'd work properly.

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/app_manager.js:
```diff
@@ -279,83 +265,106 @@ hqDefine('app_manager/js/app_manager', function () {
-            if ($sortable.find(ui.item).length < 1) {
-                return;
-            }
+            if ($sortable.find(ui.item).length < 1) { return; }
```
I abhor this style...maybe we can argue about it in book club.

esoergel replied:
which style?  doing this on one line, or short-circuiting early?

orangejenny replied:
the single line

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/app_manager.js:
```diff
@@ -232,12 +232,159 @@ hqDefine('app_manager/js/app_manager', function () {
+
+            var from, to;
+            if (movingFormToNewModule) {
+                [from, to] = calculateMoveFormToNewModule($sortable, ui, toModuleUid);
```
Huh, I haven't seen that style in js. I like the style from a coding perspective, but I'm opposed to parse errors, since they can make it harder to integrate with other tools that expect correct syntax. We've had issues with the uglifier failing on js syntax that's illegal but accepted by browsers, specifically having a trailing comma in a list of function args.

I'm curious, does the uglier work properly on this? You can test by running `build_requirejs` locally (make sure to `rm -rf staticfiles` and run `collectstatic` and `compilejsi18n` first).

esoergel replied:
Dang, I'll just switch it to basic JS, how's this look?
```js
var move;
if ... {
    move = ...;
} else {
    move = ...;
}
var from = move[0],
    to = move[1]
```

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/app_manager.js:
```diff
@@ -279,83 +265,106 @@ hqDefine('app_manager/js/app_manager', function () {
+            $sortable.children().not('.sort-disabled').each(function (i) {
+                if ($(this).data('moduleuid') !== toModuleUid) {
+                    to = i;
+                    from = parseInt(ui.item.data('index'), 10);
```
Is the `parseInt` necessary? Usually jquery's `data` will parse numbers out automatically.

esoergel replied:
Not sure, these were here already

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/app_manager.js:
```diff
@@ -279,83 +265,106 @@ hqDefine('app_manager/js/app_manager', function () {
+            });
+            return [from, to];
+        }
+        function saveDataToForm($form, from, to, fromModuleUid, toModuleUid, sortingForms) {
```
Nitpick: this name is vague. `storeDestinationAndTarget`?

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/app_manager.js:
```diff
@@ -247,7 +248,34 @@ hqDefine('app_manager/js/app_manager', function () {
+            $(".module").each(function (index, element) {
+                $(element).data('index', index);
+                modulesByUid[ $(element).data('uid') ] = element;
+                if ( $(element).data('rootmoduleid') ) {
```
Should this be `rootmoduleuid`? I just went back to check this wouldn't run into issues with rootmoduleid being zero, rootmoduleuid would make it more obvious that isn't an issue.

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/app_manager.js:
```diff
@@ -247,7 +248,34 @@ hqDefine('app_manager/js/app_manager', function () {
+        }
+        function addChildModuleToParent(childModule, parent) {
+            var childList = $(parent).find("ul.child-modules");
+            if ( childList.length === 0 ) {
```
Nitpick: Are you attached to the style of a space after opening paren/bracket and before closing paren/bracket? I feel like we don't use it much, at least in js.

orangejenny replied:
Nevermind, looks like stickler has the same complaint.

orangejenny commented on corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html:
```diff
@@ -12,11 +12,18 @@
+            <li class="edit-module-li js-sorted-li module"
+                data-index="{{ module.id }}"
+                data-uid="{{ module.unique_id }}"
+                data-rootModuleId="{{ module.root_module_id|default:"" }}"
```
Nitpick: is the `default` necessary?

esoergel replied:
Yeah, otherwise it's `"None"`

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/app_manager.js:
```diff
@@ -234,10 +234,14 @@ hqDefine('app_manager/js/app_manager', function () {
-            initSortable($(this));
-        });
-        $('.sort-action').hide();
+        if (modulesWereReordered()) {
```
quality function name

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/app_manager.js:
```diff
@@ -234,10 +234,14 @@ hqDefine('app_manager/js/app_manager', function () {
+            $('.sortable').each(function () {
+                initSortable($(this));
+            });
+            $('.sort-action').hide();
```
Nitpick: It's a good habit to use `addClass('hide')` instead of `.hide()`, because the latter directly updates an inline style, which can screw with other styling.

esoergel replied:
Actually, do you know of any reason that's not just hidden in the html?  Why bother doing that in JS?

orangejenny replied:
Hmm, I don't know. I'd definitely favor doing it in html.

orangejenny commented on corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html:
```diff
@@ -67,3 +67,35 @@
+                        {% trans "Your modules were reordered" %}
+                    </p>
+                    <p>
+                        Click below to move these child modules under their parents.
```
trans

orangejenny commented on corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html:
```diff
@@ -67,3 +67,35 @@
+                </div>
+                <div class="modal-footer">
+                    <button type="submit" class="disable-on-submit btn btn-danger">
+                        Rearrange
```
trans

orangejenny commented on corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html:
```diff
@@ -67,3 +67,35 @@
+                    </p>
+                </div>
+                <div class="modal-footer">
+                    <button type="submit" class="disable-on-submit btn btn-danger">
```
Is this a dangerous action or is this what we want them to do?

Either way I'd add an explicit cancel button so people don't feel trapped.

esoergel replied:
I am trying to force them to make that change, though I was also thinking about a "cancel", since popping this up every pageload and not allowing them to reorder modules is already a pretty aggressive way of forcing it.

orangejenny replied:
I'd mark it as `btn-primary` then.

orangejenny commented on corehq/apps/app_manager/views/apps.py:
```diff
@@ -902,6 +902,27 @@ def rearrange(request, domain, app_id, key):
+            else:
+                orphaned_modules.append(module)
+
+    app.modules = [m for modules in modules_by_parent_id.values() for m in modules]
```
Nitpick: having `modules` twice inside the list comprehension is a little confusing.

esoergel replied:
what would you suggest?  I need to join the lists that are the dict values.  Could also do:
```python
list(itertools.chain(*modules_by_parent_id.values()))
```

orangejenny replied:
I think my real complaint is I don't quite get what this function is doing unless I go look at other commits for context. We're reordering modules, but why? What order are they going to end up in? Plus I find nested list comprehensions confusing which isn't helping. (In general when reviewing I waffle over whether to criticize nested comprehensions as difficult to read or just accept they're something I need to get more comfortable with.)

I think if the overall purpose of the function were clearer I wouldn't get hung up on this line. How about a function-level comment? Or a more descriptive function name?

esoergel replied:
That's fair - I'll add in a docstring.  Nested list comprehensions took me a bit to get, but what finally did it for me is just remembering that the for statements are in the same order as in a regular nested loop.  So these two are equivalent:
```python
players = [player for team in teams for player in team]

players = []
for team in teams:
    for player in team:
        players.add(player)
```
I do think reversing the order would make more sense, but eh, that's how python works, I guess.  Just a matter of getting used to it.  I mentally move `player` to the end.

orangejenny replied:
Sounds good. Thanks for the example - I shall study ;)

orangejenny reviewed (approved):
This is fabulous. My commentary is all minor.

esoergel committed:
Use basic JS so the linter is happy

esoergel committed:
Revert "Alert user of orphaned child modules"

This reverts commit 3ba0b179f4ae9e73d5fea22d6704cbd5de8fd9d2.

esoergel committed:
Misc PR feedback

esoergel commented on corehq/apps/app_manager/models.py:
```diff
@@ -5410,14 +5410,14 @@ def rearrange_modules(self, i, j):
         self.modules = modules
 
-    def rearrange_forms(self, to_module_id, from_module_id, i, j):
+    def rearrange_forms(self, to_module_uid, from_module_uid, i, j):
```
Wow, good call, I was just updating this and totally assumed it'd be `rearrange_forms(from, to)`, but it's actually the reverse.

orangejenny replied:
Nice rename, any chance you'd mind also updating the order so it *is* `(from, to)`?

orangejenny replied:
thank you!

esoergel committed:
Use semantic names for nonlocal index variables

esoergel committed:
Update rearrange modal

esoergel commented:
Added a few commits in response to PR feedback, and updated the warning modal with some better copy.  

After looking a little more at the details of the rollout plan, I'm going to add in a feature flagged backwards compatibility shim so we don't have to migrate everyone (there are many dozens of copies of the ICDS apps floating around, and they have custom case tiles which have hardcoded module indices).

esoergel committed:
Add un-nested child-modules flag

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/app_manager.js:
```diff
@@ -326,7 +326,7 @@ hqDefine('app_manager/js/app_manager', function () {
             if (to !== from || movingFormToNewModule) {
                 var $form = $sortable.find('> .sort-action form');
-                saveDataToForm($form, from, to, fromModuleUid, toModuleUid, sortingForms);
+                storeRearrangementInForm($form, from, to, fromModuleUid, toModuleUid, sortingForms);
```
nice

esoergel committed:
Rearrange from to

esoergel committed:
Add docstring

esoergel committed:
Declare css classes up front rather than in JS

Also only init drag handles when allowing sorting

esoergel committed:
Add migration to handle existing child module apps

orangejenny commented on corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html:
```diff
@@ -100,7 +100,7 @@ <h4 class="modal-title">
                         {% trans "Not Now" %}
                     </button>
-                    <button type="submit" class="disable-on-submit btn btn-danger">
+                    <button type="submit" class="disable-on-submit btn btn-danger btn-primary">
```
remove `btn-danger`

esoergel replied:
Good call
https://github.com/dimagi/commcare-hq/pull/23676

orangejenny commented on corehq/toggles.py:
```diff
@@ -1041,6 +1041,19 @@ def _commtrackify(domain_name, toggle_is_enabled):
     [NAMESPACE_DOMAIN]
 )
 
+LEGACY_CHILD_MODULES = StaticToggle(
```
Should this and the migration be in a separate PR that's released before this one, so the legacy apps are flagged before the new code goes out?

esoergel replied:
Sorry, I thought I'd replied to this days ago.  I wanted the migration be run during the deploy since it also reorders modules in some circumstances, which I'd like to happen right when we roll out the new UI, otherwise it might introduce additional confusion.

orangejenny replied:
sounds good

orangejenny commented on corehq/apps/app_manager/migrations/0003_legacy_child_modules.py:
```diff
@@ -0,0 +1,72 @@
+
+def flag_legacy_child_module_domains(*args, **kwargs):
+    """Enable the LEGACY_CHILD_MODULES flag for domains that need it"""
+    domains = set(toggles.BASIC_CHILD_MODULE.get_enabled_domains() +
```
Should this also include apps that use custom case tiles, even if they don't currently use child modules, in case they turn child modules on in the future?

esoergel replied:
I don't think so - the two aren't incompatible, but it's not always backwards compatible.  If a project starts using child modules later, they'll only have an issue when they make a new child module which changes the module indices.  That's similar to adding a normal new module or otherwise changing the ordering, and I think its appropriate to leave it to them to address.

orangejenny replied:
cool, thanks for explaining

esoergel committed:
Use javascript directly for rearrangement

Previously this assembled an HTML form only to later serialize it
anyways

esoergel committed:
Split up form and module rearrangements

They're actually pretty different - doesn't make sense to pretend
otherwise

esoergel committed:
Module rearrangements happen with respect to all

esoergel committed:
Calculate all indices anew

This was trying to minimally reset indices, which was complicated, difficult to
reason about, and error prone.  This change much more simply traverses the
structure and resets the indices.

esoergel committed:
Simplify to/from calculation

This wasn't returning proper values in some instances, and to be honest
I didn't really understand what exactly it was trying to do.  This new
version pulls the "from" from the element itself, since we already have
it, and it calculates "to" based on the index of the element with that
UID in the target sortable.  Previously it looked for the element in the
target sortable with a different module UID.

esoergel committed:
Use module UID for comparison

Feels safer than index, and we have it anyways

esoergel commented:
This PR is getting big, and more work is coming out of QA.  Consequently, I'll no longer be adding commits directly to this PR - instead, I'll PR all changes in from additional branches.

esoergel committed:
rm btn-danger

orangejenny committed:
Merge pull request #23676 from dimagi/es/module-display-btn

Use btn-primary

esoergel committed:
Merge pull request #23675 from dimagi/es/module-display-QA

Improve form and module reordering logic

esoergel committed:
Update the sidebar when modules are (de)parented

esoergel committed:
Save new ordering

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into es/module-display

esoergel committed:
Fix moving module with children lower

Since children will also be taken out of the flow, the indices were
wrong.  This simplifies and uses indexing without children.

esoergel committed:
Be more explicit about selector scope

esoergel committed:
Merge pull request #23713 from dimagi/es/make-child-module

Update sidebar when child modules are created or changed

esoergel committed:
Use normal assignment with jsonobject

The old version of this set the modules as expected, but the ones added
via the += didn't persist on save.  Really bizzare bug - I'm guessing it
comes from jsonobject overriding the assignment operator or __iadd__ or
something.

esoergel committed:
Alert the user to delete the module's children

esoergel committed:
Merge pull request #23753 from dimagi/es/dont-delete-parents

Don't delete parent modules willy-nilly

esoergel committed:
Merge pull request #23750 from dimagi/es/add-modules

Add modules properly

esoergel committed:
Switch from automated migration to manual mgmt cmd

This command now only flags projects that need the legacy flag; it will
not modify any projects, everyone else who needs a change will see a pop-up.

Accordingly, I'm switching this to a mgmt command so I can run it in
advance of the rollout

esoergel committed:
Merge pull request #23767 from dimagi/es/rm-automatic-migration

Switch from automated migration to manual mgmt cmd

esoergel committed:
Use legacy reordering logic under legacy flag

esoergel committed:
Merge pull request #23783 from dimagi/es/legacy-sort

Use legacy reordering logic under legacy flag

esoergel committed:
Make help text explain what's happening

dannyroberts committed:
Merge branch 'master' into es/module-display

orangejenny committed:
Merge branch 'master' into es/module-display

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into es/module-display

esoergel committed:
Child menus -> submenus

esoergel committed:
Merge pull request #24611 from dimagi/es/child-module-text

Text updates for sub-menus feature
